main.go

```diff
@@ -1,64 +1,94 @@
 package main
 
 import(
-	"github.com/vocdoni/gnark-crypto-primitives/poseidon"
-	"github.com/consensys/gnark/frontend/cs/r1cs"
-	"github.com/consensys/gnark/backend/groth16"
-	"github.com/consensys/gnark-crypto/ecc"
-	"github.com/consensys/gnark/frontend"
-	"tutorial/sumcheck-verifier-circuit/polynomials"
-	// "github.com/kr/pretty"
-	"math/big"
+    "github.com/vocdoni/gnark-crypto-primitives/poseidon"
+    "github.com/consensys/gnark/frontend/cs/r1cs"
+    "github.com/consensys/gnark/backend/groth16"
+    "github.com/consensys/gnark-crypto/ecc"
+    "github.com/consensys/gnark/frontend"
+    "tutorial/sumcheck-verifier-circuit/polynomials"
+    "math/big"
 )
 
 type Circuit struct {
-	Seed frontend.Variable `gnark:"Seed"` 
-	C1 frontend.Variable `gnark:"C1"`
-	R1 frontend.Variable `gnark:"R1"`
-	R2 frontend.Variable `gnark:"R2"`
-	G_Coeffs [4]frontend.Variable `gnark:"G Coefficients"`
-	G1_Coeffs [2]frontend.Variable `gnark:"G1 Coefficients"`
-	G2_Coeffs [2]frontend.Variable `gnark:"G2 Coefficients"`
+    Seed frontend.Variable `gnark:"Seed"` 
+    C1 frontend.Variable `gnark:"C1"`
+    GCoeffs []frontend.Variable `gnark:"G Coefficients"`
+    GiCoeffs [][]frontend.Variable `gnark:"G_i Coefficients"`
+}
+
+func check01 (api frontend.API, result frontend.Variable, polyCoeffs []frontend.Variable) {
+    g0 := polynomials.UniP(polyCoeffs, 0, api)
+    g1 := polynomials.UniP(polyCoeffs, 1, api)
+    api.AssertIsEqual(result, api.Add(g0, g1))
+}
+
+func checkFirstRound (api frontend.API, circuit *Circuit) {
+    check01(api, circuit.C1, circuit.GiCoeffs[0])
+}
+
+func checkMiddleRounds (api frontend.API, circuit *Circuit, hashes *[]frontend.Variable) {
+    var poseidonInstance = poseidon.NewPoseidon(api)
```

Comment:
I think this is correct, but it would be much better if you moved the whole poseidon logic into a new struct and abstract it away. Have something like `Transcript` that holds the last hash value and an array of values to be bound and then you have an API like `Bind(frontend.Variable...)` that writes a bunch of new values and `GetChallenge() frontend.Variable` that does the hashing if necessary and returns the challenge. It will be much cleaner to show correctness of such an object.

Reply:
Abstracted away the Poseidon logic